File: src/structures/Vector2d.java
package structures;

/**
 * Representation of a two dimensional mathematical vector. Supports typical
 * operations on vectors in an immutable fashion. <li>Note that the angle of a
 * vector is defined as being measured in the counter-clockwise direction on the
 * x-axis. <li>Angles are always stored in radians.
 * @author Sean Lewis
 */

public class Vector2d {

	private double xComp;
	private double yComp;
	private double length;
	private double angle;

	/**
	 * Creates a zero vector.
	 */
	public Vector2d() {
		this(0.0, 0.0);
	}

	/**
	 * Creates a new vector by defining its perpendicular components.
	 * @param xComponent the magnitude of the horizontal component
	 * @param yComponent the magnitude of the vertical component
	 */
	public Vector2d(double xComponent, double yComponent) {
		xComp = xComponent;
		yComp = yComponent;
		calculateProperties();
	}

	/**
	 * Creates a new vector by defining its magnitude and direction
	 * @param magnitude the magnitude of the vector
	 * @param angle the angle of the vector
	 * @param dummy dummy to differentiate between constructors
	 */
	public Vector2d(double magnitude, double angle, boolean dummy) {
		length = magnitude;
		this.angle = angle;
		calculateComponenets();
	}

	/**
	 * Creates a new vector with an initial and terminal point.
	 * @param p1 the initial point
	 * @param p2 the terminal point
	 */
	public Vector2d(Point2d p1, Point2d p2) {
		xComp = p2.x - p1.x;
		yComp = p2.y - p1.y;
		calculateProperties();
	}

	private void calculateComponenets() {
		xComp = length * Math.cos(angle);
		yComp = length * Math.sin(angle);
	}

	private void calculateProperties() {
		angle = CalcHelp.getAngle(xComp, yComp);
		length = Math.hypot(xComp, yComp);
	}

	/**
	 * Returns the vector sum of this and another vector.
	 * @param v a vector
	 * @return the sum
	 */
	public Vector2d add(Vector2d v) {
		return new Vector2d(xComp + v.getXComponent(), yComp
				+ v.getYComponent());
	}

	/**
	 * Returns the vector difference of this and another vector.
	 * @param v a vector
	 * @return the difference
	 */
	public Vector2d subtract(Vector2d v) {
		return new Vector2d(xComp - v.getXComponent(), yComp
				- v.getYComponent());
	}

	/**
	 * Returns the scalar multiplication of this and a constant.
	 * @param k a constant
	 * @return the product
	 */
	public Vector2d multiply(double k) {
		return new Vector2d(xComp * k, yComp * k);
	}

	/**
	 * Returns the dot product (scalar multiplication) of this and another
	 * vector
	 * @param v a vector
	 * @return the dot product
	 */
	public double dot(Vector2d v) {
		return xComp * v.getXComponent() + yComp * v.getYComponent();
	}

	/**
	 * Returns the positive, acute angle between this and another vector.
	 * @param v a vector
	 * @return the angle between the vectors
	 */
	public double angleBetween(Vector2d v) {
		return Math.acos(dot(v) / (length * v.getLength()));
	}

	/**
	 * Returns the projection of this vector onto v
	 * @param v a vector
	 * @return the projection onto v
	 */
	public Vector2d projection(Vector2d v) {
		return v.multiply(this.dot(v) / (v.dot(v)));

	}

	/**
	 * Returns the projection of this vector in the direction of an angle
	 * @param t an angle
	 * @return the projection directed in t's direction
	 * @deprecated use the projection method with vector parameter instead
	 */
	public Vector2d projection(double t) {
		return new Vector2d(length * Math.cos(angle - t), t, true);
	}

	/**
	 * Returns the horizontal component of this vector
	 * @return the horizontal component
	 */
	public double getXComponent() {
		return xComp;
	}

	/**
	 * Sets the horizontal component of this vector
	 * @param xComponent the new horizontal component
	 */
	public void setXComponent(double xComponent) {
		this.xComp = xComponent;
		calculateProperties();
	}

	/**
	 * Returns the vertical component of this vector
	 * @return the vertical component
	 */
	public double getYComponent() {
		return yComp;
	}

	/**
	 * Sets the vertical component of this vector
	 * @param yComponent the new vertical component
	 */
	public void setYComponent(double yComponent) {
		this.yComp = yComponent;
		calculateProperties();
	}

	/**
	 * Returns the magnitude of this vector
	 * @return the magnitude
	 */
	public double getLength() {
		return length;
	}

	/**
	 * Sets the magnitude of this vector
	 * @param length the new magnitude
	 */
	public void setLength(double length) {
		this.length = length;
		calculateComponenets();
	}

	/**
	 * Returns the angle of this vector
	 * @return the angle
	 */
	public double getAngle() {
		return angle;
	}

	/**
	 * Sets the angle of this vector
	 * @param angle the new angle
	 */
	public void setAngle(double angle) {
		this.angle = angle;
		calculateComponenets();
	}

	/**
	 * Returns a mathematical description of this vector with unit vector
	 * notation.
	 */
	public String toString() {
		return "<" + xComp + "i" + ((yComp >= 0) ? (" + ") : (" - "))
				+ Math.abs(yComp) + "j>";
	}

}

File: src/structures/Vector2d.scala
package structures

/** Representation of a two dimensional mathematical vector. Supports typical
 * operations on vectors in an immutable fashion. <li>Note that the angle of a
 * vector is defined as being measured in the counter-clockwise direction on the
 * x-axis. <li>Angles are always stored in radians.
 */
class Vector2d(val xComponent: Double, val yComponent: Double) {

  /** The norm of the vector. */
  val magnitude = math.hypot(xComponent, yComponent)
  /** The angle formed by the vector with the positive x axis.*/
  val angle = CalcHelp.getAngle(xComponent, yComponent)

  /** Creates a new zero vector. */
  def this() = this(0.0, 0.0)
  
  /** Creates a vector as determined by a point. */
  def this(p: Point2d) = this(p.x, p.y)

  /** Defines a vector in polar coordinates. */
  def polarVector(magnitude: Double, angle: Double) =
    new Vector2d(magnitude * math.cos(angle), magnitude * math.sin(angle))

  /** Negates a vector. */
  def unary_- = new Vector2d(-xComponent, -yComponent)

  /** Returns the sum of two vectors. */
  def add(v: Vector2d): Vector2d = 
    new Vector2d(xComponent + v.xComponent, yComponent + v.yComponent)

  /** Returns the sum of two vectors. */
  def +(v: Vector2d): Vector2d = this.add(v)

  /** Returns the difference of two vectors. */
  def subtract(v: Vector2d): Vector2d = this + (-v)

  /** Returns the difference of two vectors. */
  def -(v: Vector2d): Vector2d = this.subtract(v)
     
  /** Returns the scalar multiple of this vector by k. */ 
  def multiply(k: Double): Vector2d =
    new Vector2d(k * xComponent, k * yComponent)

  /** Returns the scalar multiple of this vector by 1/k. */
  def divide(k: Double): Vector2d = this.multiply(1.0 / k)

  /** Returns the scalar multiple of this vector by k. */
  def *(k: Double): Vector2d = this.multiply(k)

  /** Returns the scalar multiple of this vector by 1/k. */
  def /(k: Double): Vector2d = this.multiply(1.0 / k)

  /** Returns the dot product of two vectors. */
  def dot(v: Vector2d): Double =
    xComponent * v.xComponent + yComponent * v.yComponent

  /** Returns a unit vector in the direction of this vector. */
  def normalize(): Vector2d = this / this.dot(this)

  /** Returns the projection of this vector onto v. */
  def proj(v: Vector2d): Vector2d = v * (this.dot(v) / v.dot(v))
    
  /** Returns the projection of this vector in the direction of angle. */
  def proj(angle: Double): Vector2d
    = polarVector(magnitude * math.cos(this.angle - angle), angle)
  
  /** Returns if another type can equal this Vector2d. */
  def canEqual(other: Any): Boolean = 
    other.isInstanceOf[structures.Vector2d]  
  
  override def equals(other: Any): Boolean = {
    other match {
      case that: structures.Vector2d => 
        that.canEqual(Vector2d.this) && 
        xComponent == that.xComponent &&
        yComponent == that.yComponent
      case _ => false
    }
  }

  /** Returns if another instance can equal this Vector to epsilon precision. */
  def equals(other: Any, epsilon: Double): Boolean = {
    def eq(x: Double, y: Double) = math.abs(x - y) <= epsilon
    other match {
      case that: structures.Vector2d =>
        that.canEqual(Vector2d.this) &&
          eq(xComponent, that.xComponent) &&
          eq(yComponent, that.yComponent)
      case _ => false
    }
  }
  
  override def hashCode(): Int = {
    val prime = 41
    prime * (prime + xComponent.hashCode) + yComponent.hashCode
  }

  override def toString(): String = {
    "<" + xComponent + "i" +
      (if (yComponent >= 0) (" + ") else (" - ")) +
      math.abs(yComponent) + "j>"
  }
  
}